src/esp/gfx/CubeMap.h

```diff
@@ -0,0 +1,202 @@
+   * {imageFilePrefix}.{texType}.+X.png
+   * {imageFilePrefix}.{texType}.-X.png
+   * {imageFilePrefix}.{texType}.+Y.png
+   * {imageFilePrefix}.{texType}.-Y.png
+   * {imageFilePrefix}.{texType}.+Z.png
+   * {imageFilePrefix}.{texType}.-Z.png
+   * @return true, if success, otherwise false
+   */
+  bool saveTexture(TextureType type, const std::string& imageFilePrefix);
+
+  /**
+   * ```
+   *           +----+
+   *           | -Y |
+   * +----+----+----+----+
+   * | -Z | -X | +Z | +X |
+   * +----+----+----+----+
+   *           | +Y |
+   *           +----+
+   * ```
+   * NOTE: +Y is top
+   * @brief load cubemap texture from external images
+   * @param type can be "rgba", "depth", or "objectId" (TODO)
+   * @param imageFilePrefix, the prefix of the image filename
+   * @param imageFileExtension, the image filename extension (such as "png",
+   * "jpg")
+   * @return true if succeeded, otherwise false
+   * The 6 image files then would be:
+   * {imageFilePrefix}.{texType}.+X.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.-X.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.+Y.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.-Y.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.+Z.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.-Z.{imageFileExtension}
+   */
+  void loadTexture(TextureType type,
+                   const std::string& imageFilePrefix,
+                   const std::string& imageFileExtension);
+
+  /**
+   * @brief Render to cubemap texture using the camera
+   * @param camera, a cubemap camera
+   * NOTE: It will NOT automatically generate the mipmap for the user
+   */
+  void renderToTexture(CubeMapCamera& camera,
+                       scene::SceneGraph& sceneGraph,
+                       RenderCamera::Flags flags);
+
+ protected:
+  Flags flags_;
+  int imageSize_ = 0;
+  std::map<TextureType, std::unique_ptr<Magnum::GL::CubeMapTexture>> textures_;
+
+  Magnum::ResourceManager<Magnum::Trade::AbstractImporter> imageImporterManger_;
+
+  /**
+   * @brief Recreate textures
+   */
+  void recreateTexture();
+
+  // framebuffer for drawable selection
+  Magnum::GL::Framebuffer frameBuffer_{Magnum::NoCreate};
+
+  // in case there is no need to output depth texture, we need a depth buffer
+  Magnum::GL::Renderbuffer optionalDepthBuffer_;
+
+  /**
+   * @brief Recreate frame buffer
+   */
+  void recreateFramebuffer();
+
+  /**
+   * @brief Prepare to draw to the texture
+   */
+  void prepareToDraw(int cubeSideIndex);
+
+  /**
+   * @brief Map shader output to attachments.
+   */
+  void mapForDraw();
+
+  /**
+   * @brief check if the class instance is created with corresponding texture
+   * enabled
+   */
+  void textureTypeSanityCheck(TextureType type,
+                              const std::string& functionNameStr);
+
+  /**
+   * @brief convert cube face index to Magnum::GL::CubeMapCoordinate
+   */
+  Magnum::GL::CubeMapCoordinate convertFaceIndexToCubeMapCoordinate(
```

Comment:
Very good point. Makes a lot of sense. Will change.

Reply:
> to avoid a false impression that they're needed outside of the class as well. 

Actually, they are private functions, so they cannot be called outside of the class.